registry/network/functions.go

```diff
@@ -0,0 +1,521 @@
+// and decrements the address by 1. In case of an underflow (e.g., decrementing
+// 0.0.0.0 in IPv4), an error is returned.
+//
+// Parameters:
+//
+//	ip net.IP - the IP address to decrement.
+//
+// Returns:
+//
+//	net.IP - the decremented IP address.
+//	error - an error if the IP address underflows or if the IP version cannot be determined.
+//
+// Example:
+//
+//	{{ parseIP "192.168.0.2" | ipDecrement }} // Output: 192.168.0.1
+//	{{ parseIP "ffff::1" | ipDecrement }}     // Output: ffff::
+func (nr *NetworkRegistry) IPDecrement(ip net.IP) (net.IP, error) {
+	switch nr.determineIPVersion(ip) {
+	case 4:
+		ip = ip.To4()
+	case 6:
+		ip = ip.To16()
+	}
+
+	dec := 1 // decrement value
+	for i := len(ip) - 1; i >= 0 && dec > 0; i-- {
+		if ip[i] < byte(dec) { // detect underflow
+			ip[i] = 0xFF
+		} else {
+			ip[i] -= byte(dec)
+			dec = 0 // stop further decrements
+		}
+	}
+
+	if dec > 0 {
+		return net.IP{}, errors.New("ip decrement underflow")
+	}
+
+	return ip, nil
+}
+
+// CIDRContains checks if a given IP address is contained within a specified CIDR block.
+//
+// It parses both the CIDR block and the IP address, and checks whether the IP falls
+// within the network range defined by the CIDR. If either the CIDR or the IP address is invalid,
+// an error is returned.
+//
+// Parameters:
+//
+//	cidrStr string - the string representation of the CIDR block.
+//	ip string - the string representation of the IP address to check.
+//
+// Returns:
+//
+//	bool - true if the IP address is within the CIDR block, false otherwise.
+//	error - an error if the CIDR block or IP address cannot be parsed.
+//
+// Example:
+//
+//	{{ cidrContains "192.168.0.0/24" "192.168.0.1" }}  // Output: true
+//	{{ cidrContains "192.168.0.0/24" "10.0.0.1" }}     // Output: false
+func (nr *NetworkRegistry) CIDRContains(cidrStr string, ip string) (bool, error) {
+	parsedCIRDR, err := nr.ParseCIDR(cidrStr)
+	if err != nil {
+		return false, err
+	}
+
+	parsedIP, err := nr.ParseIP(ip)
+	if err != nil {
+		return false, err
+	}
+
+	return parsedCIRDR.Contains(parsedIP), nil
+}
+
+// CIDRSize calculates the total number of IP addresses in the given CIDR block.
+// It works for both IPv4 and IPv6 CIDR blocks.
+//
+// The function parses the CIDR string, determines the IP version, and calculates the
+// size of the network range based on the prefix length.
+//
+// Parameters:
+//
+//	cidr string - the string representation of the CIDR block.
+//
+// Returns:
+//
+//	*big.Int - the total number of IP addresses in the CIDR block.
+//	error - an error if the CIDR block cannot be parsed.
+//
+// Example:
+//
+//	{{ cidrSize "192.168.0.0/24" }}   // Output: 256
+//	{{ cidrSize "2001:db8::/32" }}    // Output: 79228162514264337593543950336 (IPv6 range)
+func (nr *NetworkRegistry) CIDRSize(cidrStr string) (*big.Int, error) {
+	cidr, err := nr.ParseCIDR(cidrStr)
+	if err != nil {
+		return nil, fmt.Errorf("invalid CIDR block: %w", err)
```

Comment:
This wrapped error message is great, other CIDR method you add should have something equivalent. example CIDRContains, CIDRfirst, CIDRLast ...